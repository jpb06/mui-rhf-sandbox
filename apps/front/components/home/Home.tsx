import { Title } from '@mui/icons-material';
import { Avatar, CircularProgress, Divider, Grid, List } from '@mui/material';

import { SubTitle } from '@components/data-display';
import { FullpageBox } from '@components/layout';

import { Skill } from './children/Skill';
import { useRedirectOnNoUserData } from './hooks/useRedirectOnNoUserData';
import { stringAvatar } from './logic/avatar.logic';

export const Home = () => {
  const user = useRedirectOnNoUserData();

  if (!user) {
    return (
      <FullpageBox>
        <CircularProgress />
      </FullpageBox>
    );
  }

  return (
    <FullpageBox>
      <Grid
        container
        direction="column"
        justifyContent="center"
        alignItems="center"
      >
        <Grid item>
          <Avatar
            {...stringAvatar(`${user.firstName} ${user.lastName}`)}
            sx={{
              width: 80,
              height: 80,
              mb: 2,
              fontSize: 'xx-large',
            }}
          />
        </Grid>
      </Grid>
      <Grid item>
        <Title>{`${user.firstName} ${user.lastName}`}</Title>
      </Grid>
      <Grid>
        <SubTitle>You are a {user.role.name}</SubTitle>
      </Grid>
      <Divider
        sx={{
          width: '90%',
          mt: 2,
          mb: 2,
        }}
      />
      <Grid item>
        <SubTitle>Your skills are</SubTitle>
        <List dense>
          {user.skills.map((skill) => (
            <Skill key={skill.id} {...skill} />
          ))}
        </List>
      </Grid>
    </FullpageBox>
  );
};
