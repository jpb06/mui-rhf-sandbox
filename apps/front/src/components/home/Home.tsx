import { Avatar, CircularProgress, Divider, Grid, List } from '@mui/material';

import { Title, SubTitle, FullpageBox } from '@components';

import { Skill } from './children/Skill';
import { useRedirectOnNoUserData } from './hooks/useRedirectOnNoUserData';
import { stringAvatar } from './logic/avatar.logic';

export const Home = () => {
  const user = useRedirectOnNoUserData();

  if (!user) {
    return (
      <FullpageBox>
        <CircularProgress />
      </FullpageBox>
    );
  }

  return (
    <FullpageBox>
      <Grid
        container
        direction="column"
        justifyContent="center"
        alignItems="center"
      >
        <Grid item>
          <Avatar
            {...stringAvatar(`${user.firstName} ${user.lastName}`)}
            sx={{
              width: 80,
              height: 80,
              mb: 2,
              fontSize: 'xx-large',
            }}
          />
        </Grid>
      </Grid>
      <Grid item>
        <Title>{user.userName}</Title>
        <SubTitle>{`${user.firstName} ${user.lastName}`}</SubTitle>
      </Grid>
      <Grid>
        <SubTitle>You are a {user.role.name}</SubTitle>
      </Grid>
      <Divider
        sx={{
          width: '90%',
          mt: 2,
          mb: 2,
        }}
      />
      {user.skills.length > 0 && (
        <Grid item>
          <SubTitle>Your skills are</SubTitle>
          <List dense>
            {user.skills.map((skill) => (
              <Skill key={skill.id} {...skill} />
            ))}
          </List>
        </Grid>
      )}
    </FullpageBox>
  );
};
