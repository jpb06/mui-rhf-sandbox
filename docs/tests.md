# âš¡ Testing

Let's talk a bit about the various types of tests we can do and what problems they do solve.

What we aim to do is trophy testing, basing ourself on [Typescript](https://www.typescriptlang.org), [jest](https://jestjs.io), [testing library](https://testing-library.com), [msw](https://mswjs.io) and [cypress](https://www.cypress.io).

![Diagram](./assets/testing-trophy.jpg)

## ðŸ”¶ Unit tests

They focus on testing a module - could be a component or a function - in isolation. That means any outside context this module may rely upon will be mocked. This are pretty straightforward tests, which are typically really fast.

You can find unit tests in the [molecules or organisms folder of the shared components library](./../libs/front/components/src/molecules). Here is a simple example:

```typescript
describe('Brand component', () => {
  it('should display brand informations', () => {
    render(<Brand color="white" />);

    expect(screen.getByText('Sandbox')).toBeInTheDocument();
    expect(
      screen.getByText('nextjs / react-hook-form / testing-library')
    ).toBeInTheDocument();
  });
});
```

## ðŸ”¶ Visual regression tests

These tests check that the small bricks of our app didn't drastically change visually. This is pretty useful when we defined our own system design relying on a whole set of generic components.

## ðŸ”¶ Snapshots

Snapshots are useful to make sure we didn't visually break another part of the application by touching a generic component for example.

You can find snapshot tests in the templates folder of the front app:

- [Logged user home](./../apps/front/src/templates/logged-user-home/LoggedUserHome.spec.tsx).
- [Signup form](./../apps/front/src/templates/signup-form/SignupForm.spec.tsx).

```typescript
describe('snapshots', ()=> {
  it('should match snapshot when loading', () => {
    msw.userDataQuery(200, undefined);
  
    const { baseElement } = render(<LoggedUserHome />);
  
    expect(baseElement).toMatchSnapshot();
  });

  it('should match snapshot when displaying user data', async () => {
    msw.userDataQuery(200, mockedUser);
  
    const { baseElement } = render(<LoggedUserHome />);
  
    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    expect(baseElement).toMatchSnapshot();
  });
})
```

## ðŸ”¶ Integration tests

Integration tests have the highest return on investment because they are not as hard to write and maintain as e2e tests while giving us good confidence about a part of our system.

You can find integration tests in the templates folder:

- [Logged user home](./../apps/front/src/templates/logged-user-home/LoggedUserHome.spec.tsx).
- [Signup form](./../apps/front/src/templates/signup-form/SignupForm.spec.tsx).

```typescript
it('should display user data', async () => {
  msw.userDataQuery(200, mockedUser);
  
  render(<LoggedUserHome />);
  
  await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

  const { firstName, lastName, userName } = mockedUser;

  expect(screen.getByText(`${firstName} ${lastName}`)).toBeInTheDocument();
  expect(screen.getByText(`${userName}`)).toBeInTheDocument();
});
```

## ðŸ”¶ End to end testing

The tests giving us the most confidence. They also cost a lot. So it's generally wiser to only write e2e for key features of our application; the ones that represent a critical risk for the product.
