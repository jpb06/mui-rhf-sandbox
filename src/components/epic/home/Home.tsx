import { Avatar, CircularProgress, Divider, Grid, List } from '@mui/material';

import { FullpageBox } from '@components/generic/layout/fullpage-box/FullpageBox';
import { SubTitle } from '@components/generic/titles/SubTitle/SubTitle';
import { Title } from '@components/generic/titles/Title/Title';
import { stringAvatar } from '@logic/avatar.logic';

import { Skill } from './children/Skill';
import { useRedirectOnNoUserData } from './hooks/useRedirectOnNoUserData';

export const Home = () => {
  const user = useRedirectOnNoUserData();

  if (!user) {
    return (
      <FullpageBox>
        <CircularProgress />
      </FullpageBox>
    );
  }

  return (
    <FullpageBox>
      <Grid
        container
        direction="column"
        justifyContent="center"
        alignItems="center"
      >
        <Grid item>
          <Avatar
            {...stringAvatar(`${user.firstName} ${user.lastName}`)}
            sx={{
              width: 80,
              height: 80,
              mb: 2,
              fontSize: 'xx-large',
            }}
          />
        </Grid>
      </Grid>
      <Grid item>
        <Title>{`${user.firstName} ${user.lastName}`}</Title>
      </Grid>
      <Grid>
        <SubTitle>You are a {user.role.name}</SubTitle>
      </Grid>
      <Divider
        sx={{
          width: '90%',
          mt: 2,
          mb: 2,
        }}
      />
      <Grid item>
        <SubTitle>Your skills are</SubTitle>
        <List dense>
          {user.skills.map((skill) => (
            <Skill key={skill.id} {...skill} />
          ))}
        </List>
      </Grid>
    </FullpageBox>
  );
};
