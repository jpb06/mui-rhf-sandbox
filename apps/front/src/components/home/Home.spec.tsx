import {
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';

import { render } from '@components/tests';
import { msw, mocks, mockedData } from '@tests';

import { Home } from './Home';

jest.mock('next/router');

describe('Signup component', () => {
  const { pushMock } = mocks.nextRouter();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render vanilla html/css in snapshot', () => {
    msw.userDataQuery(200, undefined);

    const { baseElement } = render(<Home />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should display a loading indicator', async () => {
    msw.userDataQuery(200, undefined);

    render(<Home />);

    await screen.findByRole('progressbar');
  });

  it('should redirect to signup if no user data is available', async () => {
    msw.userDataQuery(200, undefined);

    render(<Home />);

    await waitFor(() => {
      expect(pushMock).toHaveBeenCalledTimes(1);
    });
  });

  it('should display user data', async () => {
    msw.userDataQuery(200, mockedData.signedUser);

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(
      `${mockedData.signedUser.firstName} ${mockedData.signedUser.lastName}`
    );
    screen.getByText(`${mockedData.signedUser.userName}`);
  });

  it('should display the communication skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 6, name: 'Communication', category: 'Soft skills' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByTestId('RecordVoiceOverIcon');
    screen.getByText(/communication/i);
  });

  it('should display the information sharing skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 8, name: 'Information sharing', category: 'Soft skills' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByTestId('QuestionAnswerIcon');
    screen.getByText(/information sharing/i);
  });

  it('should display the project drive skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [
        {
          id: 7,
          name: 'Project drive',
          category: 'Management',
        },
      ],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByTestId('TwoWheelerIcon');
    screen.getByText(/project drive/i);
  });

  it('should display the reporting skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 9, name: 'Reporting', category: 'Management' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByTestId('AssessmentIcon');
    screen.getByText(/reporting/i);
  });

  it('should display the roadmap definition skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 11, name: 'Roadmap definition', category: 'Management' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByTestId('LoyaltyIcon');
    screen.getByText(/roadmap definition/i);
  });

  it('should display the jest skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 1, name: 'jest', category: 'Tech' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByRole('img', { name: /jest/i });
    screen.getByText(/jest/i);
  });

  it('should display the react skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 4, name: 'react', category: 'Tech' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByRole('img', { name: /react/i });
    screen.getByText('react');
  });

  it('should display the typescript skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 5, name: 'Typescript', category: 'Tech' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByRole('img', { name: /typescript/i });
    screen.getByText(/typescript/i);
  });

  it('should display the github skill', async () => {
    msw.userDataQuery(200, {
      ...mockedData.signedUser,
      skills: [{ id: 10, name: 'Github actions', category: 'Tech' }],
    });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));

    screen.getByText(/your skills are/i);

    screen.getByTestId('GitHubIcon');
    screen.getByText(/github/i);
  });

  it('should not display a skills section', async () => {
    msw.userDataQuery(200, { ...mockedData.signedUser, skills: [] });

    render(<Home />);

    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'));
  });
});
