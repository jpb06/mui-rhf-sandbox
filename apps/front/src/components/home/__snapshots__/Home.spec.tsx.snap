// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component should render vanilla html/css in snapshot 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-1 {
  0% {
    stroke-dasharray: 1px,200px;
    stroke-dashoffset: 0;
  }

  50% {
    stroke-dasharray: 100px,200px;
    stroke-dashoffset: -15px;
  }

  100% {
    stroke-dasharray: 100px,200px;
    stroke-dashoffset: -125px;
  }
}

.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  min-height: 100vh;
  -webkit-background-size: cover;
  background-size: cover;
  padding: 16px;
}

.emotion-0>.MuiGrid-item {
  max-width: none;
}

.emotion-1 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  overflow: hidden;
  width: 100%;
  padding-bottom: 16px;
  background-color: rgba(255, 255, 255, 0.88);
  border-radius: 16px;
  max-width: 700px;
}

.emotion-2 {
  display: block;
  -webkit-background-size: cover;
  background-size: cover;
  background-repeat: no-repeat;
  -webkit-background-position: center;
  background-position: center;
  height: 150px;
}

.emotion-3 {
  padding: 16px;
}

.emotion-3:last-child {
  padding-bottom: 24px;
}

.emotion-4 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-bottom: 0px;
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-7 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #ff8f00;
}

.emotion-9 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 1.25rem;
  line-height: 1.6;
  letter-spacing: 0.0075em;
  color: #ff8f00;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: 12px;
  color: #0097a7;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  padding-top: 0px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-12>:not(:first-of-type) {
  margin-left: 8px;
}

.emotion-13 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-13>.MuiGrid-item {
  max-width: none;
}

.emotion-14 {
  display: inline-block;
  color: #1976d2;
  -webkit-animation: animation-0 1.4s linear infinite;
  animation: animation-0 1.4s linear infinite;
}

.emotion-15 {
  display: block;
}

.emotion-16 {
  stroke: currentColor;
  stroke-dasharray: 80px,200px;
  stroke-dashoffset: 0;
  -webkit-animation: animation-1 1.4s ease-in-out infinite;
  animation: animation-1 1.4s ease-in-out infinite;
}

<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column emotion-0"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-1"
      >
        <div
          class="MuiCardMedia-root emotion-2"
          role="img"
          style="background-image: url(/img/banner.jpg);"
        />
        <div
          class="MuiCardContent-root emotion-3"
        >
          <div
            class="MuiGrid-root MuiGrid-container emotion-4"
          >
            <div
              class="MuiGrid-root MuiGrid-container emotion-5"
            >
              <div
                class="MuiGrid-root MuiGrid-item emotion-6"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-7"
                  data-testid="BuildIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m22.7 19-9.1-9.1c.9-2.3.4-5-1.5-6.9-2-2-5-2.4-7.4-1.3L9 6 6 9 1.6 4.7C.4 7.1.9 10.1 2.9 12.1c1.9 1.9 4.6 2.4 6.9 1.5l9.1 9.1c.4.4 1 .4 1.4 0l2.3-2.3c.5-.4.5-1.1.1-1.4z"
                  />
                </svg>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item emotion-6"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 emotion-9"
                >
                  Sandbox
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container emotion-4"
            >
              <div
                class="MuiGrid-root MuiGrid-item emotion-11"
              >
                nextjs / react-hook-form / testing-library
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing emotion-12"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column emotion-13"
          >
            <span
              class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary emotion-14"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg emotion-15"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate emotion-16"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
